File: js/sort_list.js
/**
 * Sort a linked list in O(n log n) time using constant space complexity.

Example 1:

Input: 4->2->1->3
Output: 1->2->3->4
Example 2:

Input: -1->5->3->4->0
Output: -1->0->3->4->5
 */

/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var sortList = function(head) {
  if (!head) return null;
  // divide and concur
  // break list into half, need to find half point first

  // find the length of the list
  let n = 0;
  let count = head;
  while (count) {
    n += 1;
    count = count.next;
  }

  return sort(head, n);

  function sort(node, length) { 
    if (length === 1) { //consider the length === 1
      return node;
    }
    const mid = Math.ceil(length / 2);
    let lefthead = node;
    for(let i = 0; i < mid-1; i++) {
        lefthead = lefthead.next;
    }
    let righthead = lefthead.next;
    lefthead.next = null;
    let sortedLeft = sort(node, mid);    
    let sortedRight = sort(righthead, length - mid);

    return mergeSortedList(sortedLeft, sortedRight);
  }

  function mergeSortedList(l1, l2) {
    const head = new ListNode(0);
    let current = head;
    while (l1 && l2) {
      if (l1.val <= l2.val) {
        current.next = l1;
        l1 = l1.next;
        current = current.next;
      } else {
        current.next = l2;
        l2 = l2.next;
        current = current.next;
      }
    }
    if (l1) current.next = l1;
    if (l2) current.next = l2;
    return head.next;
  }
};

File: py/sortList.py
"""
Sort a linked list in O(n log n) time using constant space complexity.

Example 1:

Input: 4->2->1->3
Output: 1->2->3->4
Example 2:

Input: -1->5->3->4->0
Output: -1->0->3->4->5
"""

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution:
    def sortList(self, head: ListNode) -> ListNode:
        if not head:
            return 
        # Calulate the length
        n = 0
        curr = head
        while curr:
            n += 1
            curr = curr.next
        return self.helper(head, n)

    def helper(self, node, length):
        if length == 1:
            return node
        mid = int(length/2)
        leftHead = node
        for i in range(mid-1):
            leftHead = leftHead.next
        rightHead = leftHead.next
        leftHead.next = None
        left = self.helper(node, mid)
        right = self.helper(rightHead, length - mid)

        return self.mergeList(left, right)

    def mergeList(self, l1, l2):
        dummy = ListNode(0)
        head = dummy
        while l1 and l2:
            if l1.val > l2.val:
                head.next = ListNode(l2.val)
                l2 = l2.next
            else:
                head.next = ListNode(l1.val)
                l1 = l1.next
            head = head.next
        while l1:
            head.next = ListNode(l1.val)
            l1 = l1.next
            head = head.next
        while l2:
            head.next = ListNode(l2.val)
            l2 = l2.next
            head = head.next
        return dummy.next
